cpp/src/transform/jit/kernel.cu

```diff
@@ -66,59 +85,94 @@ struct scalar {
                                            cudf::size_type row)
   {
-    return Accessor::element(views, 0);
+    return Accessor::element(outputs, 0);
   }
 
-  static __device__ decltype(auto) element(cudf::column_device_view_core const* views,
+  static __device__ decltype(auto) element(cudf::column_device_view_core const* inputs,
                                            cudf::size_type row)
   {
-    return Accessor::element(views, 0);
+    return Accessor::element(inputs, 0);
   }
 
-  static __device__ void assign(cudf::mutable_column_device_view_core const* views,
+  static __device__ void assign(cudf::mutable_column_device_view_core const* outputs,
                                 cudf::size_type row,
                                 type value)
   {
-    return Accessor::assign(views, 0, value);
+    return Accessor::assign(outputs, 0, value);
   }
 };
 
-template <typename Out, typename... In>
-CUDF_KERNEL void kernel(cudf::mutable_column_device_view_core const* output,
-                        cudf::column_device_view_core const* inputs)
+template <bool has_user_data, typename Out, typename... In>
+CUDF_KERNEL void kernel(cudf::mutable_column_device_view_core const* outputs,
+                        cudf::column_device_view_core const* inputs,
+                        void* user_data)
 {
+  // inputs to JITIFY kernels have to be either sized-integral types or pointers. Structs or
+  // references can't be passed directly/correctly as they will be crossing an ABI boundary
+
   // cannot use global_thread_id utility due to a JIT build issue by including
   // the `cudf/detail/utilities/cuda.cuh` header
   auto const block_size          = static_cast<thread_index_type>(blockDim.x);
   thread_index_type const start  = threadIdx.x + blockIdx.x * block_size;
   thread_index_type const stride = block_size * gridDim.x;
-  thread_index_type const size   = output->size();
+  thread_index_type const size   = outputs[0].size();
 
   for (auto i = start; i < size; i += stride) {
-    GENERIC_TRANSFORM_OP(&Out::element(output, i), In::element(inputs, i)...);
+    if constexpr (has_user_data) {
+      GENERIC_TRANSFORM_OP(user_data, i, &Out::element(outputs, i), In::element(inputs, i)...);
+    } else {
+      GENERIC_TRANSFORM_OP(&Out::element(outputs, i), In::element(inputs, i)...);
+    }
   }
 }
 
-template <typename Out, typename... In>
-CUDF_KERNEL void fixed_point_kernel(cudf::mutable_column_device_view_core const* output,
-                                    cudf::column_device_view_core const* inputs)
+template <bool has_user_data, typename Out, typename... In>
+CUDF_KERNEL void fixed_point_kernel(cudf::mutable_column_device_view_core const* outputs,
+                                    cudf::column_device_view_core const* inputs,
+                                    void* user_data)
 {
   // cannot use global_thread_id utility due to a JIT build issue by including
   // the `cudf/detail/utilities/cuda.cuh` header
```

Comment:
Can we work on this TODO in a future PR? Should be a matter of splitting up that header.

Reply:
Yes, I already created an issue for it https://github.com/rapidsai/cudf/issues/18216, I'll get to it when I can, it's only a few hours of work but it's not a priority at the moment